### PR TITLE
Mets transformer pulls mets from S3 and creates invisible work

### PR DESCRIPTION
**Pull request by alicefuzier**
Part of https://github.com/wellcometrust/platform/issues/3921
Supersedes #214 
Resolves Transform METS files to works

### CODE REVIEW

File: pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/MetsData.scala [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package uk.ac.wellcome.platform.transformer.mets
+
+case class MetsData(
+  recordIdentifier: String,
+  accessCondition: Option[String]
+) {
+
+  def toWork(version: Int) = ???
```
Comment (warrd) [inline pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/MetsData.scala]: Didn't know you could do that!

File: pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/transformer/Mets.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        sourceIdentifier = sourceIdentifier,
+        workData(unidentifiableItem)
+      )
+
+  private def workData(unidentifiableItem: MaybeDisplayable[Item]) =
+    WorkData(
+      items = List(unidentifiableItem),
+      mergeCandidates = List(mergeCandidate))
+
+  private def mergeCandidate = MergeCandidate(
```
Comment (jamesgorrie) [inline pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/transformer/Mets.scala]: Does it make a difference if things are `val`s or `def`s?

File: pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/parsers/MetsXmlParser.scala [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      )
+  }
+
+  private def recordIdentifier(root: Elem): Try[String] = {
+    val identifierNodes =
+      (root \\ "dmdSec" \ "mdWrap" \\ "recordInfo" \ "recordIdentifier").toList
+    identifierNodes match {
+      case List(identifierNode) => Success(identifierNode.text)
+      case _ =>
+        Failure(new Exception("Could not parse recordIdentifier from METS XML"))
```
Comment (jamesgorrie) [inline pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/parsers/MetsXmlParser.scala]: When these errors are throw in `process` - they won't have any identifiable information with them will there?

Could we add some? I can't remember what we get from the adapter - it's the S3 path?

File: pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/transformer/Mets.scala (hunk 2 above)
Comment (warrd) [reply pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/transformer/Mets.scala]: In these sort of cases when no mutability / side effects:

`val` = Non lazy, stored in memory
`lazy val` = Lazy, stored in memory
`def` = Lazy, not stored in memory (recalculated each time needed)

Stylistically I quite like `def` here but don't think it matters much

File: pipeline/transformer/transformer_mets/src/test/scala/uk/ac/wellcome/platform/transformer/mets/fixtures/MetsGenerators.scala [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+       |          <mods:recordInfo>
+       |            <mods:recordIdentifier source="gbv-ppn">$recordIdentifier</mods:recordIdentifier>
+       |          </mods:recordInfo>
+       |          <mods:accessCondition type="dz">${license.id.toUpperCase}</mods:accessCondition>
+       |        </mods:mods>
+       |      </mets:xmlData>
+       |    </mets:mdWrap>
+       |  </mets:dmdSec>
+       |</mets:mets>
+       |""".stripMargin
```
Comment (warrd) [inline pipeline/transformer/transformer_mets/src/test/scala/uk/ac/wellcome/platform/transformer/mets/fixtures/MetsGenerators.scala]: Scala has XML literals, so could use that here. Going to be removed at some time in the future in Scala 3 AFAIK.

File: pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/parsers/MetsXmlParser.scala [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+  private def accessCondition(root: Elem): Try[Option[String]] = {
+    val licenseNodes = (root \\ "dmdSec" \ "mdWrap" \\ "accessCondition")
+      .filter(_ \@ "type" == "dz")
+      .toList
+    licenseNodes match {
+      case Nil               => Success(None)
+      case List(licenseNode) => Success(Some(licenseNode.text))
+      case _ =>
+        Failure(new Exception("Found multiple accessCondtions in METS XML"))
```
Comment (warrd) [inline pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/parsers/MetsXmlParser.scala]: In most of the codebase it seems like `Either[_, _]` are used in preference to `Try`. Do you reckon worth being consistent? (at some point might be great to have a style guide for these sort of decisions)

File: common/internal_model/src/main/resources/identifier-schemes.csv [hunk 6]
```diff
@@ -14,3 +14,4 @@ preservica-guid,Perservica GUID
 calm-authority-code,CALM authority code
 isbn,International Standard Book Number
 issn,ISSN
+mets,METS
```
Comment (warrd) [inline common/internal_model/src/main/resources/identifier-schemes.csv]: I remember some talk about whether adding a new identifier type was the correct thing. Has @jtweed got any opinion on this?

File: pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/transformer/Mets.scala (hunk 2 above)
Comment (jamesgorrie) [reply pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/transformer/Mets.scala]: I suppose my concern would be that if it's a `def` it feels like we've said `There's a reason why this recalculates each time` is all.

File: pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/parsers/MetsXmlParser.scala (hunk 5 above)
Comment (jamesgorrie) [reply pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/parsers/MetsXmlParser.scala]: > have a style guide for these sort of decisions

👍 

This function is actually getting it - so we could use `getAccessCondition` which will either get an optional license or a known error state of `multiple access conditions in METS`.

This error feels like it would be good to have somewhere associated with the b number so if asked, "why isn't this b number getting an access statement" we could point to our imaginary Success / Failure dashboard and it will be there.

Saying this - we might have (excuse the verbosity)
```scala
case class MetsAdapterFailure
case class KnownFailure extends MetsAdapterFailure
case class UnknownFailure extends MetsAdapterFailure // Probably similar to Try.Failure(err)
case class MultipleAccessConditionsInMets extends KnownFailure
def getAccessCondition(root: Elem): Either[MultipleAccessConditionsInMets, Option[AccessCondition]]
```

And then at the edge of the application you would have
```
// MetsAdapter.scala
// ...
def process: Either[MetsAdapter, UnidentifiableWork]

// Main.scala
// ...
val id = getId(input)
process(input) match {
  case Right(work) => complete(work) // continue
  case Left(err: KnownFailure) => info(err) // continue
  case Left(err: UnknownFailure) => error(err) // retry
}
```

There's some states I am missing and logic that doesn't fit right (`UnknownFailure`?).

In short I think I am saying if we know what the problem is, is it an error, or part of the application state that we know how to handle, which `Either` feels better at semantically capturing?

If this makes sense in trying to capture I can have a go at implementing it on one of the services?

Thanks to the [Scala docs](https://www.scala-lang.org/api/2.11.4/index.html#scala.util.Try) for providing clear guidance.

> The Try type represents a computation that may either result in an exception, or return a successfully computed value. It's similar to, but semantically different from the scala.util.Either type.

File: pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/transformer/Mets.scala (hunk 2 above)
Comment (alicefuzier) [reply pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/transformer/Mets.scala]: No, no reason. I tried replacing with val and it still works. The only thing is that, if they are vals, they need to be in a specific order so that each one is initialised when it's referenced by another one which is slightly annoying

File: common/internal_model/src/main/resources/identifier-schemes.csv (hunk 6 above)
Comment (alicefuzier) [reply common/internal_model/src/main/resources/identifier-schemes.csv]: We discussed this with @jamesgorrie and I think we sort of reached the conclusion that just adding this here was OK?

File: pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/parsers/MetsXmlParser.scala (hunk 5 above)
Comment (alicefuzier) [reply pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/parsers/MetsXmlParser.scala]: Yep, I'm going to switch to `Either`.

As for the Known vs Unknown failure thing, I'm not sure... There is already a [`RecognisedFailure`](https://github.com/wellcometrust/scala-messaging/blob/master/messaging/src/main/scala/uk/ac/wellcome/messaging/sqs/RecognisedFailure.scala) class which is basically used in [`SQSStream`](https://github.com/wellcometrust/scala-messaging/blob/master/messaging/src/main/scala/uk/ac/wellcome/messaging/sqs/SQSStream.scala) to log and send metrics differently, so we can extend that if we just want to change logging.

I wouldn't base the logic around retries on whether an error is known or unknown though. I would say that anything that needs action from us and that we want to be notified about should end up in the dlq eventually,  so it shouldn't be deleted from the queue, which means it should be retried. 
I.e. the multiple access conditions in mets error: it's a data error so a retry won't change the outcome, but we make it retry so that we can see it in the dlq. We can then notify an error in the data to whoever is responsible for fixing it.

File: pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/parsers/MetsXmlParser.scala (hunk 3 above)
Comment (alicefuzier) [reply pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/parsers/MetsXmlParser.scala]: How about 

> ERROR u.a.w.p.t.m.s.MetsTransformerWorkerService - There was an error processing MetsData(mets.xml,6): blah blah blah..
